ui/button.js

```diff
@@ -56,355 +39,267 @@ $.widget( "ui.button", {
+				this.element.val( this.options.label );
+			} else {
+				this.element.html( this.options.label );
+			}
+		}
+		if ( this.options.icon ) {
+			this._updateIcon( this.options.icon )._updateTooltip();
+		}
 	},
 
-	_determineButtonType: function() {
-		var ancestor, labelSelector, checked;
+	_updateTooltip: function() {
+		this.title = this.element.attr( "title" );
 
-		if ( this.element.is("[type=checkbox]") ) {
-			this.type = "checkbox";
-		} else if ( this.element.is("[type=radio]") ) {
-			this.type = "radio";
-		} else if ( this.element.is("input") ) {
-			this.type = "input";
-		} else {
-			this.type = "button";
+		if ( !this.options.showLabel && !this.title ) {
+			this.element.attr( "title", this.options.label );
 		}
+	},
 
-		if ( this.type === "checkbox" || this.type === "radio" ) {
-			// we don't search against the document in case the element
-			// is disconnected from the DOM
-			ancestor = this.element.parents().last();
-			labelSelector = "label[for='" + this.element.attr("id") + "']";
-			this.buttonElement = ancestor.find( labelSelector );
-			if ( !this.buttonElement.length ) {
-				ancestor = ancestor.length ? ancestor.siblings() : this.element.siblings();
-				this.buttonElement = ancestor.filter( labelSelector );
-				if ( !this.buttonElement.length ) {
-					this.buttonElement = ancestor.find( labelSelector );
-				}
-			}
-			this.element.addClass( "ui-helper-hidden-accessible" );
+	_updateIcon: function( icon ) {
+		if ( !this.icon ) {
+			this.icon = $( "<span>" );
+			this._addClass( this.icon, "ui-button-icon", " ui-icon" );
 
-			checked = this.element.is( ":checked" );
-			if ( checked ) {
-				this.buttonElement.addClass( "ui-state-active" );
+			if ( !this.options.showLabel ) {
+				this._addClass( "ui-button-icon-only" );
+			} else {
+				this._addClass( null, "ui-icon-" + this.options.iconPosition );
 			}
-			this.buttonElement.prop( "aria-pressed", checked );
 		} else {
-			this.buttonElement = this.element;
+			this._removeClass( this.icon, null, this.options.icon );
 		}
-	},
-
-	widget: function() {
-		return this.buttonElement;
+		this._addClass( this.icon, null, icon );
+		this.icon.appendTo( this.element );
+		return this;
 	},
 
 	_destroy: function() {
-		this.element
-			.removeClass( "ui-helper-hidden-accessible" );
-		this.buttonElement
-			.removeClass( baseClasses + " ui-state-active " + typeClasses )
-			.removeAttr( "role" )
-			.removeAttr( "aria-pressed" )
-			.html( this.buttonElement.find(".ui-button-text").html() );
+		this.element.removeAttr( "role" );
 
+		if ( this.icon ) {
+			this.icon.remove();
+		}
 		if ( !this.hasTitle ) {
-			this.buttonElement.removeAttr( "title" );
+			this.element.removeAttr( "title" );
 		}
 	},
 
 	_setOption: function( key, value ) {
+		if ( key === "icon" ) {
+			if ( value !== null ) {
+				this._updateIcon( value );
+			} else {
+				this.icon.remove();
+				this._removeClass( "ui-button-icon", " ui-icon-" + this.options.iconPosition );
```

Comment:
Also this is wrong, since "ui-button-icon" is added to the icon itself, not the button, so it can't be removed from the button. I think this should be:
```js
this._removeClass( "ui-button-icon-only", "ui-icon-" + this.options.iconPosition );
```
Or maybe those two need to be removed separately?